packages/SwingSet/src/vats/comms/clist-kernel.js

```diff
@@ -1,145 +1,145 @@
+        const p = state.promiseTable.get(lref);
+        // We should always know about this lref. It is allocated upon
+        // receipt.  We retain the promiseTable entry even after the
+        // promise is resolved (to remember the resolution).
+        assert(p, X`how did I forget about ${lref}`);
+        kfref = allocateKernelPromiseID();
+        if (!p.resolved && !deciderIsKernel(lref)) {
+          subscribeKernelToPromise(lref);
+        }
+      } else {
+        assert.fail(X`unknown type ${type}`);
       }
-      return vpid;
+      state.toKernel.set(lref, kfref);
+      state.fromKernel.set(kfref, lref);
+      cdebug(`comms add mapping l->k ${kfref}<=>${lref}`);
     }
-    assert.fail(X`cannot give type ${type} to kernel`);
+    return state.toKernel.get(lref);
   }
 
-  function provideKernelForLocalResult(vpid) {
-    if (!vpid) {
+  function provideKernelForLocalResult(lpid) {
+    if (!lpid) {
       return null;
     }
-    const p = state.promiseTable.get(vpid);
-    assert(!p.resolved, X`result ${vpid} is already resolved`);
+    const p = state.promiseTable.get(lpid);
+    assert(!p.resolved, X`result ${lpid} is already resolved`);
     // TODO: reject somehow rather than crashing weirdly if we are not
     // already the decider, but I'm not sure how we could hit that case.
-    changeDeciderToKernel(vpid);
+    changeDeciderToKernel(lpid);
     // if we knew about this promise already, and thought the kernel
     // cared.. well, it doesn't now
-    unsubscribeKernelFromPromise(vpid);
-    return vpid;
+    unsubscribeKernelFromPromise(lpid);
+    return provideKernelForLocal(lpid);
+  }
+
+  function kernelSlotToLocalSlot(kernelSlot) {
+    assert.typeof(kernelSlot, 'string', 'non-string kernelSlot');
+    if (!state.fromKernel.has(kernelSlot)) {
+      const { type } = parseVatSlot(kernelSlot);
+      let localSlot;
+      if (type === 'object') {
+        localSlot = allocateLocalObjectID();
+      } else if (type === 'promise') {
+        localSlot = allocateLocalPromiseID();
+      } else {
+        assert.fail(X`unknown type ${type}`);
+      }
+      state.fromKernel.set(kernelSlot, localSlot);
+      state.toKernel.set(localSlot, kernelSlot);
+      cdebug(`comms add mapping k->l ${kernelSlot}<=>${localSlot}`);
+    }
+    return state.fromKernel.get(kernelSlot);
+  }
+
+  function retireKernelPromiseID(kfpid) {
+    insistVatType('promise', kfpid);
+    const lpid = state.fromKernel.get(kfpid);
+    assert(lpid, X`unknown kernel promise ${kfpid}`);
+    assert(state.toKernel.has(lpid), X`unmapped local promise ${lpid}`);
+    state.fromKernel.delete(kfpid);
+    state.toKernel.delete(lpid);
+    cdebug(`comms delete mapping l<->k ${kfpid}<=>${lpid}`);
   }
 
   // *-LocalForKernel: kernel sending in to comms vat
 
-  function provideLocalForKernel(kref, doNotSubscribeSet) {
-    const { type, allocatedByVat } = parseVatSlot(kref);
+  function provideLocalForKernel(kfref, doNotSubscribeSet) {
```

Comment:
Yes, the whole `provideLocalForKernel`/`provideLocalForKernelResult` vs. `kernelSlotToLocalSlot` thing was confusing to me, but I just took it as a given and tried to make the new stuff work with minimal adjustment.  This is a place where the analogy to the kernel broke down, because the kernel has nothing corresponding to the `provideX` vs. `provideXResult` distinction we have here.  I'm not 100% convinced this distinction is even necessary, but I didn't feel compelled to diagnose whether or not that was actually true so I just rolled with what was already there.  We should discuss this 1:1.